File: cpp/199.cpp
#include <vector>
#include <algorithm>
using namespace std;

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
  vector<int> rightSideView(TreeNode *root) {
    if (root == nullptr)
      return {};
    vector<int> res{root->val};
    vector<int> l = rightSideView(root->left), r = rightSideView(root->right);
    copy(r.begin(), r.end(), back_inserter(res));
    if(l.size() > r.size())
      copy(l.begin() + r.size(), l.end(), back_inserter(res));
    return res;
  }
};

File: python/199.py
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution(object):
    def rightSideView(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        if root is None:
            return []
        l = self.rightSideView(root.left)
        r = self.rightSideView(root.right)
        return [root.val] + r + l[len(r):]

t1 = TreeNode(1)
t2 = TreeNode(2)
t1.left = t2
print(Solution().rightSideView(t1))
